[new] [jsx-curly-brace-presence] Disallow unnecessary JSX expressions or enforce them

jackyho112 opened the pull request:
For issue https://github.com/yannickcr/eslint-plugin-react/issues/1310:

This rule checks for unnecessary JSX expressions when literals alone will do or enforce the presence of JSX expressions.

References: 
[String literals in JSX](https://facebook.github.io/react/docs/jsx-in-depth.html#string-literals)
[JSX gotchas](https://github.com/facebook/react/blob/v15.4.0-rc.3/docs/docs/02.3-jsx-gotchas.md#html-entities)

ljharb commented on lib/rules/jsx-curly-brace-presence.js:
```diff
@@ -24,22 +27,20 @@ module.exports = {
-          },
-          additionalProperties: false
-        }
+          props: {enum: optionValues},
```
this should probably still specify the default?

ljharb commented on lib/rules/jsx-curly-brace-presence.js:
```diff
@@ -24,22 +27,20 @@ module.exports = {
-          additionalProperties: false
-        }
+          props: {enum: optionValues},
+          children: {enum: optionValues}
```
also here?

ljharb commented on lib/rules/jsx-curly-brace-presence.js:
```diff
@@ -108,7 +109,7 @@ module.exports = {
 
     return {
-      JSXExpressionContainer: node => {
+      JSXExpressionContainer: function(node) {
```
any particular reason to change these from arrows?

ljharb commented on lib/rules/jsx-curly-brace-presence.js:
```diff
@@ -119,7 +120,7 @@ module.exports = {
       },
 
-      Literal: node => {
+      Literal: function(node) {
```
also here?

jackyho112 commented on lib/rules/jsx-curly-brace-presence.js:
```diff
@@ -27,8 +27,8 @@ module.exports = {
         properties: {
-          props: {enum: optionValues},
-          children: {enum: optionValues}
+          props: {enum: optionValues, default: optionNever},
```
I removed the default by mistake.

jackyho112 commented on lib/rules/jsx-curly-brace-presence.js:
```diff
@@ -116,15 +116,15 @@ module.exports = {
 
     return {
-      JSXExpressionContainer: function(node) {
+      JSXExpressionContainer: node => {
```
I made the change to maintain consistencies across files, but I guess it doesn't matter anyway. I have changed it back to using the arrow function.

jackyho112 commented:
@ljharb 
Well, I hope my attempt wasn't too embarrassing. I am not sure why the travis build failed. All the tests passed for me. Let me know if you have any feedback to improve my PR.

jackyho112 commented:
@ljharb Good suggestions! I will make the changes.

jackyho112 commented:
@ljharb 
I added tests fro template literal and spread props. I also accounted for template literal in the rule and added a new rule option to allow passing in a single string to change the default. Let me know if there is any more change you have in mind. Thanks!

jseminck commented on lib/rules/jsx-curly-brace-presence.js:
```diff
@@ -0,0 +1,147 @@
+
+const optionAlways = 'always';
+const optionNever = 'never';
+const optionIgnore = 'ignore';
```
Should these constants not be all upper case and use `_` as delimiter between words, to be more consistent with the rest of the code base? See `no-danger`, `no-typos`, `no-escaped-entities`, and others...

So: `OPTION_ALWAYS`, `OPTION_NEVER` and `OPTION_IGNORE`

jackyho112 commented:
@jseminck They should. I will make the change.

jackyho112 commented:
@ljharb I added the option to fix for enforcing curly braces and had to account for a lot of edge cases. Hopefully, I haven't missed anything. Let me know if there is any more improvement you have in mind for my branch. Thanks!

jackyho112 commented:
Hey @ljharb, all checks have passed now. Let me know if you have any more change in mind for the PR and I will get them done as soon as possible. Thanks!

ljharb commented on docs/rules/jsx-curly-brace-presence.md:
```diff
@@ -0,0 +1,153 @@
+They are `always`, `always,single`, `always,double`, `always,orignal`, `never` and `ignore` for checking on JSX props and children.
+
+* `always`: always enforce curly braces inside JSX props or/and children and fix with double quotes inside the generated JSX expressions
+* `always,single`: always enforce curly braces and fix with single quotes
```
so, first - i think rather than squashing the quote style into the string value, this should be an object.

However, quote style is something that another rule should handle. This one should just pick a default - i'd go with double quotes - and let the jsx-quotes rule handle correcting it.

ljharb commented on docs/rules/jsx-curly-brace-presence.md:
````diff
@@ -0,0 +1,143 @@
+
+```js
+...
+"react/forbid-elements": [<enabled>, { "props": <string>, "children": <string> }]
````
this might be the wrong rule name

ljharb commented on docs/rules/jsx-curly-brace-presence.md:
````diff
@@ -0,0 +1,143 @@
+It can fixed to:
+
+```jsx
+<App prop='foo'>Hello world</App>;
````
let's use multiple props in this example so we can show that it allows either single or double quotes (one of each)

jackyho112 commented:
@ljharb 

Thanks for all your suggestions! They all make sense and give me an idea on how to better contribute to this repo for future PRs. Hopefully, everything looks good now!

ljharb commented on docs/rules/jsx-curly-brace-presence.md:
````diff
@@ -0,0 +1,148 @@
+
+```jsx
+<App>Hello world</App>;
+<App prop='Hello world' attr="foo" />;
````
should this example use double quotes consistently?

ljharb commented on docs/rules/jsx-curly-brace-presence.md:
````diff
@@ -0,0 +1,148 @@
+
+```jsx
+<App>{"Hello world"}</App>;
+<App prop={'Hello world'}>{'Hello world'}</App>;
````
should both of these examples use the same quote style?

ljharb commented on docs/rules/jsx-curly-brace-presence.md:
````diff
@@ -0,0 +1,148 @@
+They can be fixed to:
+```jsx
+<App>{"Hello world"}</App>;
+<App prop={'Hello world'} attr={"foo"}>{"Hello world"}</App>;
````
these should probably both use the same kind of quotes

ljharb commented on docs/rules/jsx-curly-brace-presence.md:
````diff
@@ -0,0 +1,148 @@
+It can fixed to:
+
+```jsx
+<App prop='foo' attr="bar">Hello world</App>;
````
these should probably both use the same kind of quotes

ljharb commented on docs/rules/jsx-curly-brace-presence.md:
```diff
@@ -0,0 +1,148 @@
+
+## Edge cases
+
+The fix also deals with template literals, strings with quotes and strings with escapes characters.
```
please use the Oxford comma - in this case, after "quotes"

ljharb commented on docs/rules/jsx-curly-brace-presence.md:
````diff
@@ -0,0 +1,148 @@
+will warned and fixed to:
+
+```jsx
+<App prop={'Hello "foo" world'}>{"Hello 'foo' \"bar\" world"}</App>;
````
is it expected that the quote style is unchanged for `prop`?

jackyho112 commented:
@ljharb 

Thanks again for all your suggestions. I was leaving original quotes in as much as I could previously when I was fixing the styles, but your suggestion makes sense for the sake of consistency. As a result, I have changed the operations to fix with double quotes in all cases. Hope things look better now!

jackyho112 commented:
@ljharb Thank you for the review!
